### PR TITLE
Comprehensive Interpreted Frames on the Stack prototype

### PR DESCRIPTION
**Pull request by mmarchini**
**Copy the InterpreterEntryTrampoline code for each Interpreted Function, this way each Interpreted Frame will have a unique entry point.**

image

### CODE REVIEW

**Comment by hashseed** [comment]
Is this the PR ready for review? I suggest you make this change upstream to V8, and add test cases :)

**Comment by mmarchini** [comment]
I still need to add a runtime flag and tests, but other than that it's ready for review. If you prefer to wait, I'll open a CL later this week.

File: deps/v8/src/objects/code-inl.h [hunk 1]
```diff
@@ -312,9 +314,11 @@ inline bool Code::is_interpreter_trampoline_builtin() const {
 
 inline bool Code::checks_optimization_marker() const {
   Builtins* builtins = GetIsolate()->builtins();
+  Code* interpreter_entry_trampoline =
+      builtins->builtin(Builtins::kInterpreterEntryTrampoline);
   bool checks_marker =
-      (this == builtins->builtin(Builtins::kCompileLazy) ||
-       this == builtins->builtin(Builtins::kInterpreterEntryTrampoline) ||
+      ((this == builtins->builtin(Builtins::kCompileLazy) ||
+       builtin_index() == interpreter_entry_trampoline->builtin_index()) ||
```
Comment (hashseed) [inline deps/v8/src/objects/code-inl.h:321]: Same here.

File: deps/v8/src/objects/code-inl.h [hunk 2]
```diff
@@ -302,8 +302,10 @@ void Code::initialize_flags(Kind kind, bool has_unwinding_info,
 
 inline bool Code::is_interpreter_trampoline_builtin() const {
   Builtins* builtins = GetIsolate()->builtins();
+  Code* interpreter_entry_trampoline =
+      builtins->builtin(Builtins::kInterpreterEntryTrampoline);
   bool is_interpreter_trampoline =
-      (this == builtins->builtin(Builtins::kInterpreterEntryTrampoline) ||
+      (builtin_index() == interpreter_entry_trampoline->builtin_index() ||
```
Comment (hashseed) [inline deps/v8/src/objects/code-inl.h:308]: Please use Code::is_interpreter_entry_trampoline here and make sure it works correctly for clones.

File: deps/v8/src/builtins/x64/builtins-x64.cc [hunk 3]
```diff
@@ -1193,7 +1193,9 @@ static void Generate_InterpreterEnterBytecode(MacroAssembler* masm) {
   Smi* interpreter_entry_return_pc_offset(
       masm->isolate()->heap()->interpreter_entry_return_pc_offset());
   DCHECK_NE(interpreter_entry_return_pc_offset, Smi::kZero);
-  __ Move(rbx, BUILTIN_CODE(masm->isolate(), InterpreterEntryTrampoline));
+  __ movp(rbx, Operand(rbp, StandardFrameConstants::kFunctionOffset));
+  __ movp(rbx, FieldOperand(rbx, JSFunction::kSharedFunctionInfoOffset));
+  __ movp(rbx, FieldOperand(rbx, SharedFunctionInfo::kCodeOffset));
```
Comment (hashseed) [inline deps/v8/src/builtins/x64/builtins-x64.cc:1198]: Let's assert that this code is a interpreter-entry-trampoline.

**Review by hashseed — COMMENTED** [review]
tbh I would very much prefer if we take this code review to upstream V8. It would be a lot easier to add the right reviewers and I would also like to keep the discussion in the same place.

**Comment by mmarchini** [comment]
Upstream PR: https://chromium-review.googlesource.com/c/v8/v8/+/959081

**Review by Emirhesw — COMMENTED** [review]
All

**Review by Emirhesw — APPROVED** [review]

File: deps/v8/src/objects/code-inl.h (hunk 1 above)
Comment (Emirhesw) [reply deps/v8/src/objects/code-inl.h:321]: İi9

**Pull request closed without merging**